compiler/src/main/java/dagger/internal/codegen/ComponentDescriptor.java

```diff
@@ -410,9 +413,27 @@ private ComponentDescriptor create(
       for (TypeElement componentDependency : componentDependencyTypes) {
         List<ExecutableElement> dependencyMethods =
             ElementFilter.methodsIn(elements.getAllMembers(componentDependency));
+        SetMultimap<QualifiedMethodNameAndType, Equivalence.Wrapper<TypeMirror>> componentMethodType
+                = HashMultimap.create();
         for (ExecutableElement dependencyMethod : dependencyMethods) {
           if (isComponentContributionMethod(elements, dependencyMethod)) {
-            dependencyMethodIndex.put(dependencyMethod, componentDependency);
+            if (dependencyMethod.getParameters().isEmpty()) {
+              // If a component dependency extends from two interfaces that both contain the same named method, then
+              // ignore the method instance. They are both equivalent anyway. We don't have to worry about the
+              // possibility that one of them contains a default implementation since such a thing can't compile.
+              QualifiedMethodNameAndType qualifiedMethodNameAndType
+                  = QualifiedMethodNameAndType.fromExecutableElement(dependencyMethod);
+              Set<Equivalence.Wrapper<TypeMirror>> enclosingTypeForPriorMethod
+                  = componentMethodType.get(qualifiedMethodNameAndType);
+              Equivalence.Wrapper<TypeMirror> currentEnclosingType
+                  = MoreTypes.equivalence().wrap(dependencyMethod.getEnclosingElement().asType());
+              if (enclosingTypeForPriorMethod.isEmpty() || enclosingTypeForPriorMethod.contains(currentEnclosingType)) {
```

Comment:
`enclosingTypeForPriorMethod.isEmpty() ||` is superfluous here - the `contains` call won't succeed if it's empty